### ISSUE TITLE
RecursionError: maximum recursion depth exceeded in comparison

### ISSUE THREAD

**Issue by xy-always** [ClashLuke/MinRETRO#1]
Many thanks to the author for implementing the code! May I ask if the function of the Dataset class in main.py is to encode the input sentence, but according to the paper description, isn’t the position of the token using relative encoding? Besides that, I keep getting "RecursionError: maximum recursion depth exceeded in comparison" when I run main.py. Also, could you please give me a sample of the data?

**Comment by ClashLuke** [comment]
Thanks for reporting that error.\
The issue stems from the dataset trying to find another random element that matches the minimal sequence length, as padding usually isn't wanted in autoregressive training. Instead of throwing a RecursionError, I've now opted for a cleaner `retries` option, which throws a ValueError if it can't find a "correct" element.

Unfortunately, those weren't the only issues. However, I fixed all other code that stopped the model from training.

I achieved a loss of approximately zero using a [small toy dataset](https://gist.github.com/ClashLuke/bed977a9dae3b8444a2054faf50b2221) within 2 minutes on my laptop.
![grafik](https://user-images.githubusercontent.com/39779310/202274927-be7aea8f-a741-42ab-91c6-61ba737a86c0.png)

To reproduce these results, launch main.py with the following flags: `--embedding_model_name "google/bert_uncased_L-2_H-128_A-2" --encoder-features 256 --encoder-depth 8 --encoder-heads 4 --decoder-features 256 --decoder-depth 8 --decoder-heads 4 --retrieval_frequency 1 --first_retrieve_at_depth --query-chunk-size 16 --corpus-chunk-size 32 --topk 2  --dropout-rate 0 --learning-rate 0.0002 --weight-decay 0 --batch-size 8 --sequence-length 64 --train-dataset-path data.jsonl --dataset-path data.jsonl`
